internals.Rmd

```diff
@@ -2,70 +2,558 @@
+  geom_smooth(method = "lm", formula = y ~ x) + 
+  facet_wrap(vars(year)) + 
+  ggtitle("A plot for expository purposes")
+```
+
+The reason ggplot2 is designed this way is to allow the user to continue to add new elements to a plot at a later point, without needing to recalculate anything. One implication of this is that if you want to understand the mechanics of ggplot2, you have to follow your plot as it goes down the `plot()`[^plot-note] rabbit hole. You can inspect the print method for ggplot objects by typing `ggplot2:::plot.ggplot` at the console, but for this chapter I will work with a simplified version. Stripped to its bare essentials, the ggplot2 plot method has the same structure as the following `ggprint()` function:
+
+[^plot-note]: You usually don't call this `plot()` method directly as it is invoked by the print method and thus called whenever a ggplot object is printed. 
 
 ```{r}
-ggplot2:::plot.ggplot
+ggprint <- function(x) {
+  set_last_plot(x)
+  
+  data <- ggplot_build(x)
+  gtable <- ggplot_gtable(data)
+  
+  grid::grid.newpage()
+  grid::grid.draw(gtable)
+  
+  return(invisible(x))
+}
 ```
 
-Many of the function calls above may seem foreign, and most are not really relevant to understanding ggplot2. The calls of interest are `set_last_plot(x)` (to some extent), `data <- ggplot_build(x)`, and `gtable <- ggplot_gtable(data)`. The `set_last_plot(x)` stores the plot internally so that it is retrivable with `last_plot()`. The two remaining calls are what makes up the rendering stack of ggplot2. `ggplot_build` is where the data for each layer is prepared for plotting (with everything that entails) and `ggplot_gtable` takes the prepared data and turns it into graphic elements stored in a gtable (we'll come back to what that is later). What may come as a surprise is that ggplot2 itself does not do any actual drawing. It's responsibility stops after the gtable object has been created. The gtable package which implements the gtable class does not do any drawing either. Drawing is performed by the grid package in unison with the active graphic device. This is an important point, as it means that ggplot2 and, by extension, any extensions to ggplot2, do not need to concern themselves with the nitty gritty of creating the visual output. The responsibility is solemnly on converting the user data to one or more graphic primitives such as polygons, lines, points, etc. While it is thus stricktly not true, we will continue to call this conversion into graphic primitives the *rendering process*.
+This function does not handle every possible use case, but it is sufficient to draw the plot specified above:
+
+`r columns(1, 2/3, max_width = .8)`
+```{r, cache.vars=p}
+ggprint(p) 
+```
+
+The code in our simplified print method reveals five distinct steps:
+
+- First, it calls `set_last_plot()` to store the plot object internally. This is done to allow the `last_plot()` function to work, but otherwise does not affect the plotting process. 
```

Comment:
I wonder if we should just omit `set_last_plot()` here since it's not that important?